### PR TITLE
dictionary compression for serialized types

### PR DESCRIPTION
**Pull request by normen662**

### CODE REVIEW

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto [hunk 1]
```diff
@@ -91,9 +91,10 @@ message PType {
       optional int32 field_index = 3;
     }
 
-    optional string name = 1;  // referential name -- not used in the planner
-    optional bool is_nullable = 2;
-    repeated PField fields = 3;
+    optional int32 reference_id = 1;
+    optional string name = 2;  // referential name -- not used in the planner
+    optional bool is_nullable = 3;
+    repeated PField fields = 4;
```
Comment (hatyo) [inline fdb-record-layer-core/src/main/proto/record_query_plan.proto:97]: I suppose it is ok that the field numbers are redefined considering that plan serialization is still under development, right? in other words, we will not deserialize older messages with the (now inconsistent) field numbers.

**Review by hatyo — APPROVED** [review]
LGTM

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto (hunk 1 above)
Comment (normen662) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:97]: That's right!

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanSerializationContext.java [hunk 2]
```diff
@@ -24,20 +24,23 @@
...
 
 import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
 import java.util.Objects;
 
 /**
  * Context class defining the state of serialization/deserialization currently in-flight.
  */
 @API(API.Status.INTERNAL)
 public class PlanSerializationContext {
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanSerializationContext.java:43]: done

**Review by hatyo — APPROVED** [review]
LGTM

**Pull request merged**